### PR TITLE
Add GetSchemaTable API

### PR DESCRIPTION
**Pull request by elemount**
Related with issue #307 Support GetShemaTable API.

### CODE REVIEW

File: src/MySqlConnector/Serialization/CharacterSet.cs [hunk 1]
```diff
@@ -273,4 +273,213 @@ internal enum CharacterSet : ushort
 		Utf8Mb4VietnameseUca900AccentSensitiveCaseSensitive = 300,
 		Utf8Mb4JapaneseUca900AccentSensitiveCaseSensitive = 303,
 	}
+
+	internal static partial class Extensions
+	{
+		public static int MaxLength(this CharacterSet character)
```
Comment (bgrainger) [inline src/MySqlConnector/Serialization/CharacterSet.cs]: Can you merge this into the existing `SerializationUtility.GetBytesPerCharacter`?

File: src/MySqlConnector/MySqlClient/MySqlDataReader.cs [hunk 2]
```diff
@@ -270,6 +279,100 @@ protected override void Dispose(bool disposing)
...
+				if (col.ColumnType.IsDecimal())
+				{
+					schemaRow[precision] = col.ColumnLength - 2 + ((col.ColumnFlags & ColumnFlags.Unsigned) != 0 ? 1 : 0);
+					schemaRow[scale] = col.Decimals;
+				}
+
+				schemaRow[baseSchemaName] = col.SchemaName;
+				schemaRow[baseTableName] = col.PhysicalTable;
+				schemaRow[baseColumnName] = col.PhysicalName;
+				schemaRow[isUnique] = false;
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlDataReader.cs]: Should this be set from `ColumnFlags.UniqueKey`? It doesn't appear to be covered in the tests to verify that this is the same value as Connector/NET.

File: src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs [hunk 3]
```diff
@@ -35,17 +51,90 @@ public static ColumnDefinitionPayload Create(PayloadData payload)
...
+				PhysicalTable = physicalTable,
+				PhysicalName = physicalName,
+				Decimals = decimals
+			};
+
 		}
 
-		private ColumnDefinitionPayload(string name, CharacterSet characterSet, int columnLength, ColumnType columnType, ColumnFlags columnFlags)
+		public Type GetDataType(
+			bool treadTinyAsBoolean = false,
```
Comment (bgrainger) [inline src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs]: Typo: `treatTinyAsBoolean`

File: tests/SideBySide/QueryTests.cs [hunk 4]
```diff
@@ -795,6 +795,32 @@ public void ReturnDerivedTypes()
...
+		[Fact]
+		public void GetSchemaTable()
+		{
+			var csb = AppConfig.CreateConnectionStringBuilder();
+			using (var connection = new MySqlConnection(csb.ConnectionString))
+			{
+				connection.Open();
+				using (MySqlCommand command = connection.CreateCommand())
+				{
+					command.CommandText = "select Host, ssl_cipher, max_connections, plugin, authentication_string from mysql.user limit 1";
```
Comment (bgrainger) [inline tests/SideBySide/QueryTests.cs]: If this were moved to DataTypesTests, it could test everything set up in `DataTypesFixture` (https://github.com/mysql-net/MySqlConnector/blob/master/tests/SideBySide/DataTypesFixture.cs), which should provide very broad coverage of all the data types.

File: src/MySqlConnector/MySqlClient/MySqlDataReader.cs [hunk 5]
```diff
@@ -270,6 +279,100 @@ protected override void Dispose(bool disposing)
...
+			columns.Add(isHidden);
+			columns.Add(isLong);
+			columns.Add(isReadOnly);
+
+			for (var i = 0; i < colDefinitions.Length; ++i)
+			{
+				var col = colDefinitions[i];
+				var schemaRow = schemaTable.NewRow();
+				schemaRow[columnName] = col.Name;
+				schemaRow[ordinal] = i;
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlDataReader.cs]: This is incompatible with Connector/NET which [uses 1-based ordinals](https://github.com/mysql/mysql-connector-net/blob/5864e6b21a8b32f5154b53d1610278abb3cb1cee/Source/MySql.Data/extensions/NonRT/MySqlDataReader.cs#L1000) for some reason. That seems like a bug to me, and not sure if we should clone that behaviour for backwards compatibility.

File: src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs (hunk 3 above)
Comment (bgrainger) [reply src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs]: Seems to be copied/duplicated from `ResultSet.GetFieldType`; that method should probably be refactored to delegate to this one.

File: src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs [hunk 6]
```diff
@@ -35,17 +51,90 @@ public static ColumnDefinitionPayload Create(PayloadData payload)
...
+				case ColumnType.TinyBlob:
+				case ColumnType.Blob:
+				case ColumnType.MediumBlob:
+				case ColumnType.LongBlob:
+					if (CharacterSet == CharacterSet.Binary && oldGuids && ColumnLength == 16)
+					{
+						return typeof(Guid);
+					}
+
+					return typeof(string);
```
Comment (bgrainger) [inline src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs]: Seems like this should return `typeof(byte[])` when `CharacterSet == CharacterSet.Binary`? Please increase test coverage.

**Comment by bgrainger** [comment]
Here's the skeleton of some sample tests that could be added to `DataTypes.cs`:

```
#if !NETCOREAPP1_1_2
		[Theory]
		[InlineData("rowid", "datatypes_bools", 11, typeof(int), false, true, false, 0, 0)]
		[InlineData("Boolean", "datatypes_bools", 1, typeof(bool), false, false, true, 0, 0)]
		[InlineData("SByte", "datatypes_integers", 4, typeof(sbyte), false, false, true, 0, 0)]
		[InlineData("Int16", "datatypes_integers", 6, typeof(short), false, false, true, 0, 0)]
		[InlineData("UInt16", "datatypes_integers", 5, typeof(ushort), false, false, true, 0, 0)]
		[InlineData("Double", "datatypes_reals", 22, typeof(double), false, false, true, 0, 31)]
		[InlineData("MediumDecimal", "datatypes_reals", 30, typeof(decimal), false, false, true, 28, 8)]
		[InlineData("utf8", "datatypes_strings", 300, typeof(string), false, false, true, 0, 0)]
		[InlineData("guid", "datatypes_strings", 144, typeof(Guid), false, false, true, 0, 0)]
		[InlineData("guidbin", "datatypes_strings", 144, typeof(Guid), false, false, true, 0, 0)]
		[InlineData("Blob", "datatypes_blobs", 16383, typeof(byte[]), true, false, true, 0, 0)] // 16383 is almost certainly a bug; should be 65535
		[InlineData("Date", "datatypes_times", 10, typeof(DateTime), false, false, true, 0, 0)]
		[InlineData("Time", "datatypes_times", 17, typeof(TimeSpan), false, false, true, 0, 6)]
		public void GetSchemaTable(string column, string table, int columnSize, Type dataType, bool isLong, bool isKey, bool allowDbNull, int precision, int scale)
		{
			using (var command = m_database.Connection.CreateCommand())
			{
				command.CommandText = $"select 1, `{column}` from `{table}`;";
				using (var reader = command.ExecuteReader())
				{
					var schema = reader.GetSchemaTable().Rows[1];
					Assert.Equal(column, schema["ColumnName"]);
#if BASELINE
					int ordinal = 2; // https://bugs.mysql.com/bug.php?id=61477
#else
					int ordinal = 1;
#endif
					Assert.Equal(ordinal, schema["ColumnOrdinal"]);
					Assert.Equal(columnSize, schema["ColumnSize"]);
					Assert.Equal(dataType, schema["DataType"]);
					Assert.Equal(isLong, schema["IsLong"]);
					Assert.Equal(isKey, schema["IsKey"]);
					Assert.Equal(allowDbNull, schema["AllowDBNull"]);
					Assert.Equal(precision, schema["NumericPrecision"]);
					Assert.Equal(scale, schema["NumericScale"]);
					Assert.Equal(m_database.Connection.Database, schema["BaseSchemaName"]);
					Assert.Equal(table, schema["BaseTableName"]);
					Assert.Equal(column, schema["BaseColumnName"]);
					Assert.False((bool) schema["IsUnique"]);
					Assert.False((bool) schema["IsRowVersion"]);
					Assert.False((bool) schema["IsReadOnly"]);
				}
			}
		}
#endif
```

They currently pass against Connector/NET, but not this PR. Each difference would need to be investigated to determine whether this implementation is an improvement (e.g., starting at 0-based columns, which fixes https://bugs.mysql.com/bug.php?id=61477) or a regression.

We need to have a compelling rationale for each incompatibility with Connector/NET.

**Comment by bgrainger** [comment]
(And the tests should be expanded to cover every possible data type, not just the dozen or so I put in there as an example.)

File: src/MySqlConnector/Serialization/CharacterSet.cs (hunk 1 above)
Comment (bgrainger) [reply src/MySqlConnector/Serialization/CharacterSet.cs]: I've pushed a change to `SerializationUtility.GetBytesPerCharacter` so this should be removed in favour of that: 83c1cea6f22ac4547e157a2eb93fccc85e6850a5

File: src/MySqlConnector/MySqlClient/MySqlDataReader.cs (hunk 2 above)
Comment (elemount) [reply src/MySqlConnector/MySqlClient/MySqlDataReader.cs]: Oracle set UniqueKey to be `false` always, what's your idea on this backwards compatibility?

File: src/MySqlConnector/MySqlClient/MySqlDataReader.cs (hunk 5 above)
Comment (elemount) [reply src/MySqlConnector/MySqlClient/MySqlDataReader.cs]: Let me fix it for backwards compatibility.

File: src/MySqlConnector/Serialization/CharacterSet.cs (hunk 1 above)
Comment (elemount) [reply src/MySqlConnector/Serialization/CharacterSet.cs]: Sure.

File: src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs (hunk 3 above)
Comment (elemount) [reply src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs]: Fixed by use `ResultSet.GetFieldType`

File: src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs (hunk 6 above)
Comment (elemount) [reply src/MySqlConnector/Serialization/ColumnDefinitionPayload.cs]: Fixed by use `ResultSet.GetFieldType`

File: tests/SideBySide/QueryTests.cs (hunk 4 above)
Comment (elemount) [reply tests/SideBySide/QueryTests.cs]: Fixed.

File: src/MySqlConnector/MySqlClient/MySqlDataReader.cs (hunk 2 above)
Comment (bgrainger) [reply src/MySqlConnector/MySqlClient/MySqlDataReader.cs]: I'm OK with keeping it hardcoded to `false`; it's easier and backwards-compatible.

File: tests/SideBySide/DataTypesFixture.cs [hunk 7]
```diff
@@ -121,8 +121,8 @@ public DataTypesFixture()
   latin1 varchar(300) character set 'latin1' null,
   latin1bin varchar(300) character set latin1 collate latin1_bin null,
   cp1251 varchar(300) character set 'cp1251' null,
-  guid char(36) null,
-  guidbin char(36) binary null
+  guid char(36) character set 'utf8mb4' null,
+  guidbin char(36) binary character set 'utf8mb4' null
```
Comment (bgrainger) [inline tests/SideBySide/DataTypesFixture.cs]: Why was this change needed?

Comment (bgrainger) [reply tests/SideBySide/DataTypesFixture.cs]: OK, I see that the "baseline" code returns 108 for this column length (== 36 * 3, so probably defaulting to `utf8`?). IMO, this is a bug in Oracle's connector; the `ColumnSize` of a `char(36)` column should be reported as 36, no matter the encoding. 

MariaDB's failure is interesting... I wonder if you could work around them both for now by using `character set 'ascii'` (and an expected column size of 36)?

File: src/MySqlConnector/MySqlClient/MySqlConnectionStringBuilder.cs [hunk 8]
```diff
@@ -431,7 +431,7 @@ static MySqlConnectionStringOption()
 
 			AddOption(UseCompression = new MySqlConnectionStringOption<bool>(
 				keys: new[] { "Compress", "Use Compression", "UseCompression" },
-				defaultValue: false));
+				defaultValue: true));
```
Comment (bgrainger) [inline src/MySqlConnector/MySqlClient/MySqlConnectionStringBuilder.cs]: ?

File: tests/SideBySide/AppConfig.cs [hunk 9]
```diff
@@ -34,7 +34,7 @@ public static class AppConfig
 
 		public static ServerFeatures SupportedFeatures => (ServerFeatures) Enum.Parse(typeof(ServerFeatures), Config.GetValue<string>("Data:SupportedFeatures"));
 
-		public static bool SupportsJson => SupportedFeatures.HasFlag(ServerFeatures.Json);
+		public static bool SupportsJson => false;
```
Comment (bgrainger) [inline tests/SideBySide/AppConfig.cs]: ?

File: src/MySqlConnector/MySqlClient/MySqlConnectionStringBuilder.cs (hunk 8 above)
Comment (elemount) [reply src/MySqlConnector/MySqlClient/MySqlConnectionStringBuilder.cs]: false commit, please ignore it.

File: tests/SideBySide/AppConfig.cs (hunk 9 above)
Comment (elemount) [reply tests/SideBySide/AppConfig.cs]: False commit, please ignore it.

File: tests/SideBySide/DataTypesFixture.cs (hunk 7 above)
Comment (elemount) [reply tests/SideBySide/DataTypesFixture.cs]: I read the code and it is a special case for GUID. I decide to do something special for GUID length which is different with Orcale's connector, what's your idea?

File: tests/SideBySide/DataTypes.cs [hunk 10]
```diff
@@ -624,8 +624,8 @@ public void GetSchemaTable(string column, string table, int columnSize, Type dat
 					Assert.Equal(column, schema["ColumnName"]);
 					int ordinal = 2; // https://bugs.mysql.com/bug.php?id=61477
 					Assert.Equal(ordinal, schema["ColumnOrdinal"]);
-					Assert.Equal(columnSize, schema["ColumnSize"]);
 					Assert.Equal(dataType, schema["DataType"]);
+					Assert.Equal(columnSize, schema["ColumnSize"]);
```
Comment (bgrainger) [inline tests/SideBySide/DataTypes.cs]: Appveyor runs the tests against this library, then against MySql.Data (to check for backwards compatibility). In this case, it would be OK to add a `#if BASELINE` check for the cases we know Oracle's connector has bugs (GUIDs and BLOBs).

**Comment by bgrainger** [comment]
Don't worry about codacy "failure"; I was interested in seeing what it could do, so I added it to this project (but didn't realise it would analyse every current PR).
